PyMI/src/wmi/__init__.py

```diff
@@ -754,21 +757,30 @@ def modify_instance(self, instance, operation_options=None):
 
     @mi_to_wmi_exception
     @avoid_blocking_call
-    def delete_instance(self, instance, operation_options=None):
-        # Deleting an instance using WMIDCOM fails with
-        # "Provider is not capable of the attempted operation"
-        if self._protocol != mi.PROTOCOL_WINRM:
-            tmp_session = self._app.create_session(
-                computer_name=self._computer_name,
-                protocol=mi.PROTOCOL_WINRM,
-                destination_options=self._destination_options)
-        else:
-            tmp_session = self._session
-
+    def _delete_instance(self, session, instance, operation_options=None):
         operation_options = self._get_mi_operation_options(
             operation_options=operation_options)
-        tmp_session.delete_instance(self._ns, instance._instance,
-                                    operation_options)
+        session.delete_instance(self._ns, instance._instance,
+                                operation_options)
+
+    def delete_instance(self, instance, operation_options=None):
+        try:
+            self._delete_instance(self._session, instance,
+                                  operation_options)
+        except x_wmi as exc:
+            # Deleting an instance using WMIDCOM can fail with
+            # WBEM_E_PROVIDER_NOT_CAPABLE.
+            # One affected WMI class is root/cimv2:WT_Host, there may
+            # be others as well.
+            err = ctypes.c_uint(exc.com_error.hresult).value
```

Comment:
you can check for protocol first, as you win the conversation time if the protocol is already winrm:

if (self._protocol != mi.PROTOCOL_WINRM and ctypes.c_uint(exc.com_error.hresult).value == WBEM_E_PROVIDER_NOT_CAPABLE)

or you can do a try/except on the ctypes.c_uint(exc.com_error.hresult).value transformation, as that part maybe it can fail too?
                    ):

Reply:
About the check order, the overhead is so small that it's negligible. About failing to extract the hresult, that shouldn't fail, unless the exception is incorrectly constructed, in which case we may just propagate the exception.